crates/iceberg/src/spec/transform.rs

```diff
@@ -0,0 +1,766 @@
+    ///
+    /// - Source type could be any type..
+    /// - Return type is Source type.
+    Void,
+}
+
+impl Transform {
+    /// Get the return type of transform given the input type.
+    /// Returns `None` if it can't be transformed.
+    pub fn result_type(&self, input_type: &Type) -> Option<Type> {
+        match self {
+            Transform::Identity => {
+                if matches!(input_type, Type::Primitive(_)) {
+                    Some(input_type.clone())
+                } else {
+                    None
+                }
+            }
+            Transform::Void => Some(input_type.clone()),
+            Transform::Bucket(_) => {
+                if let Type::Primitive(p) = input_type {
+                    match p {
+                        PrimitiveType::Int
+                        | PrimitiveType::Long
+                        | PrimitiveType::Decimal { .. }
+                        | PrimitiveType::Date
+                        | PrimitiveType::Time
+                        | PrimitiveType::Timestamp
+                        | PrimitiveType::Timestamptz
+                        | PrimitiveType::String
+                        | PrimitiveType::Uuid
+                        | PrimitiveType::Fixed(_)
+                        | PrimitiveType::Binary => Some(Type::Primitive(PrimitiveType::Int)),
+                        _ => None,
+                    }
+                } else {
+                    None
+                }
+            }
+            Transform::Truncate(_) => {
+                if let Type::Primitive(p) = input_type {
+                    match p {
+                        PrimitiveType::Int
+                        | PrimitiveType::Long
+                        | PrimitiveType::String
+                        | PrimitiveType::Binary
+                        | PrimitiveType::Decimal { .. } => Some(input_type.clone()),
+                        _ => None,
+                    }
+                } else {
+                    None
+                }
+            }
+            Transform::Year | Transform::Month | Transform::Day => {
+                if let Type::Primitive(p) = input_type {
+                    match p {
+                        PrimitiveType::Timestamp
+                        | PrimitiveType::Timestamptz
+                        | PrimitiveType::Date => Some(Type::Primitive(PrimitiveType::Int)),
+                        _ => None,
+                    }
+                } else {
+                    None
+                }
+            }
+            Transform::Hour => {
+                if let Type::Primitive(p) = input_type {
+                    match p {
+                        PrimitiveType::Timestamp | PrimitiveType::Timestamptz => {
+                            Some(Type::Primitive(PrimitiveType::Int))
+                        }
+                        _ => None,
+                    }
+                } else {
+                    None
+                }
+            }
+        }
```

Comment:
In Java and Python we also have the UnknownTransform. It doesn't happen much, but people could use their custom transform. Or a transform that is not yet supported by `iceberg-rust`, such as a [multi-arg bucket transform](https://github.com/apache/iceberg/pull/8259). In this case we would fall back to an unknown transform, and we cannot take any advantage of partitioning (we fall back to a full table scan 😭)

Reply:
Thanks for writing. I was aware of `UnknownTransform` but had no idea why it exists, thanks for the explaination.